modules/ding_gatewayf/ding_gatewayf.module

```diff
@@ -0,0 +1,458 @@
+}
+
+/**
+ * Redirect the user after login.
+ *
+ * If destination query string exists the user is redirect to it else send to
+ * profile page.
+ *
+ * @param string $url
+ *   The url to redirect to if destination is not set in the request. Defaults
+ *   to /user.
+ */
+function _ding_gatewayf_redirect_user($url = 'user') {
+  if (!empty($_REQUEST['destination'])) {
+    // When logged in with gatewayf the destination is part of the get request.
+    // So to use the cleaned up destination we unset it.
+    unset($_GET['destination']);
+
+    // We do not user drupal_get_destination here as it will set the destination
+    // if one do not exists.
+    $url = $_REQUEST['destination'];
+  }
+
+  drupal_goto($url);
+}
+
+/**
+ * Get user credentials from SAML attributes and the user provider.
+ *
+ * @param array $attributes
+ *   WAYF attributes.
+ *
+ * @return array
+ *   Attributes filter based on the active provider. Defaults to
+ *   'schacPersonalUniqueID' with social security number (CPR).
+ */
+function _ding_gatewayf_get_credentials(array $attributes) {
+  // Set default required attributes.
+  $required_attributes = array(
+    'schacPersonalUniqueID' => array(
+      'field' => 'CPR',
+      'authname' => TRUE,
+    ),
+  );
+
+  // Ask the provider about required attributes and give it a change to override
+  // defaults.
+  if (ding_provider_implements('wayf', 'attributes')) {
+    $required_attributes = ding_provider_invoke('wayf', 'attributes', $required_attributes);
+  }
+
+  // Filter the attributes from WAYF/SAML and only return required once.
+  return _ding_gatewayf_get_required_attributes($attributes, $required_attributes);
+}
+
+/**
+ * Extract required attributes from available WAYF attributes.
+ *
+ * @param array $attributes
+ *   WAYF attributes.
+ * @param array $required_attributes
+ *   Required attributes.
+ *
+ * @return array
+ *   Will be empty if not all required attributes is present.
+ */
+function _ding_gatewayf_get_required_attributes(array $attributes, array $required_attributes) {
+  $attribute_list = array();
+
+  if (isset($attributes) && is_array($required_attributes)) {
+    foreach ($required_attributes as $attribute => $value) {
+      if (isset($attributes[$attribute])) {
+        $attribute_list[$attribute] = _ding_gatewayf_extract_data($attributes[$attribute], $value['field']);
+      }
+    }
+
+  }
+
+  return $attribute_list;
+}
+
+/**
+ * Extracts a field from a SAML attribute string.
+ *
+ * @param string $data
+ *   The attribute form the authentication provider.
+ * @param string $field
+ *   The field to extract from the string.
+ *
+ * @return string
+ *   The value of the field.
+ */
+function _ding_gatewayf_extract_data($data, $field = '') {
+  if (empty($field)) {
+    return $data[0];
+  }
+  else {
+    $field .= DING_WAYF_DK_LOGIN_FIELD_DELIMITER;
+
+    if (($pos = strpos($data[0], $field)) !== FALSE) {
+      return substr($data[0], $pos + strlen($field));
```

Comment:
Should this use [`drupal_substr`](https://api.drupal.org/api/drupal/includes%21common.inc/group/php_wrappers/7) etc.?